### PR TITLE
Add exponential backoff retry to transaction retry decorator

### LINKED ISSUE #22: Retry delay
Would it make sense to add an optional delay or backoff to the retry decorator?  
Or would the recommendation be to use https://github.com/nameko/nameko-amqp-retry in this case?

### PR DESCRIPTION
**Pull request by daviskirk**
* Adds arguments `retries` and `backoff_factor` to retry_transaction decorator
* Default behavior stays the same
* Closes #22

### CODE REVIEW

File: nameko_sqlalchemy/transaction_retry.py [hunk 1]
```diff
@@ -1,28 +1,61 @@
...
+            try:
+                return wrapped(*args, **kwargs)
+            except exceptions:
+                errors += 1
+
+                if errors > retries:
+                    raise
+
+                if errors >= 2:
+                    backoff_value = backoff_factor * (2 ** (errors - 2))
```
Comment (mattbennett) [inline nameko_sqlalchemy/transaction_retry.py]: First retry is always immediate. Second retry is always after `backoff_factor` seconds. Subsequent retry delays are the `backoff_factor` multiplied by an increasing power of two. Is that correct?

File: nameko_sqlalchemy/transaction_retry.py [hunk 2]
```diff
@@ -1,28 +1,61 @@
...
+                    elif session:
+                        session.rollback()
                 raise
 
+        return run_or_rollback()
+
     return wrapper(wrapped)  # pylint: disable=E1120
+
+
+def retry(retries, backoff_factor, exceptions):
```
Comment (mattbennett) [inline nameko_sqlalchemy/transaction_retry.py]: Nameko has a retry decorator utility too: https://github.com/nameko/nameko/blob/master/nameko/utils/retry.py

The backoff algorithm is different, but you could use that instead?

File: nameko_sqlalchemy/transaction_retry.py (hunk 1 above)
Comment (daviskirk) [reply nameko_sqlalchemy/transaction_retry.py]: yes, correct, the algorithm is a standard exponential backoff.  In essence this implementation should be no different than the algorithm nameko uses with the exception of the initial "immediate" retry (see below).

File: nameko_sqlalchemy/transaction_retry.py (hunk 2 above)
Comment (daviskirk) [reply nameko_sqlalchemy/transaction_retry.py]: Yes, definitely.  One note on mulling this over: Using the "normal" retry decorator there is no way to have an immediate retry other than setting the "delay" parameter to 0, in which case there is no backoff at all (just keep retrying immediately indefinitely).  In the use case here of database connection errors, the immediate retry is important in the pure "reconnection" case, which is probably 95% of cases, so it kind of makes sense to start off with an immediate retry in this case.  However, I don't think this is worth having two different behaviors in the retries.

File: test/test_transaction_retry.py [hunk 3]
```diff
@@ -299,3 +300,100 @@ def test_create_with_worker_scoped_session(
...
+        (10, 0.5, [_op_exc(), ValueError(), 3], ValueError, [0]),
+        # Specified exception + multiple, then unspecified exception
+        (
+            10,
+            0.5,
+            [_op_exc() for _ in range(4)] + [5],
+            5,
+            [0.0, 0.5, 1.0, 2.0]
+        ),
+        (
```
[the remark this answers is not in the document]
Comment (daviskirk) [reply test/test_transaction_retry.py]: Corrected this, simplified the tests as the retry decorator is already tested in nameko

File: README.rst [hunk 4]
```diff
@@ -170,6 +170,18 @@ or using with the ``Database`` dependency provider
             self.db.session.add(ExampleModel(data='hello'))
             self.db.session.commit()
 
+Optionally,the transaction may be retried multiple times with an exponential backoff delay.
+In the following code, the session will be retried after 0, 1, 2 and 4 seconds, finally raising the error if the connection still has not recovered.
+
```
[the remark this answers is not in the document]
Comment (daviskirk) [reply README.rst]: Added a description of how the the backoff works.

File: nameko_sqlalchemy/transaction_retry.py (hunk 2 above)
Comment (daviskirk) [reply nameko_sqlalchemy/transaction_retry.py]: No problem, as you saw, I was having the same debate in my head.

I reverted back but added a bit of documentation.  In addition, the api and algorithm is now the same as used in the urllib3/requests retry algorithm requests: total, backoff_factor, backoff_max (https://urllib3.readthedocs.io/en/latest/reference/urllib3.util.html#urllib3.util.retry.Retry) which should be a good base for expected behavior.

**Comment by daviskirk** [comment]
@mattbennett is there anything else I can do here?

**Pull request merged**